### PR TITLE
Add subproject for headless game server

### PR DESCRIPTION
**Pull request by ssoloff**
## Overview

This PR begins the process of migrating the headless code into its own subproject: `game-headless`.  (Presumably, in the future, there will be a `game-headed` subproject that contains only the headed code, while all common code remains in `game-core`.)

I marked this PR **&lt;PREVIEW>** because, if we go forward with this split, there is one big consequence that everyone needs to understand and agree upon before it is merged:

:warning: **Users will no longer be able to run a bot from the standard installers.** :warning:

That is, if a user wants to run a bot, they will have to download an additional artifact.  The install4j installers and the portable build will _eventually_ no longer contain the code required to run a bot.

(@prastle, your opinion here would be appreciated since you have the most experience with users running their own bots.)

#### Migration plan

* Once this PR is merged, update the infrastructure scripts to use the new build artifact instead of the portable build  (_*-all_platforms.zip_).
* Once the new infrastructure scripts have been merged, we'll be clear to start migrating code to the `game-headless` project.

#### Other considerations

* We'll have an extra artifact in each release that is pretty much the same size as the  portable build (currently ~24 MiB).  Obviously, that size will decrease over time as headed code is moved out of `game-core` into its own subproject.  Likewise, the headed artifacts will decrease in size as code is moved to `game-headless`.
* We currently have to pull some files from `game-core` into the build artifact (_game_engine.properties_ and the _assets_ folder).  We'll eventually need to think of a proper way to share things like this that need to land in both the headed and headless build artifacts (hopefully we can figure out a way to decouple the headless code from the _assets_ folder).  Not sure what that will be, but for now there are a few smells in the `game-headless` buildscript that reference `project(:game-core)` directly.

## Functional Changes

None.

## Manual Testing Performed

* Verified I could run a bot from my IDE using the Eclipse launch configuration.  Tested lobby connection and starting a new game.
* Verified I could run a bot using the new build artifact.  Tested lobby connection and starting a new game.
* Created a release branch in my fork to run the complete build and verified the new artifact is built as expected:
    game-headless-release

### CODE REVIEW

**Comment by prastle** [comment]
@ssoloff  no one has been able to run a headless bot since prior to march 2016 without an additional install.  I am probably the last peson that can or does. NP Hit er

**Comment by DanVanAtta** [comment]
The migration plan sounds good. Spending a few moments on naming I think will help us. 

These are all reasonable names:

- game-headless
- headless-server
- bot-server
- triplea-bot


Similar: 
- game-headed
- triplea-client
- swing-client

From the list, any better alternatives? `sanjuro-lobby` and `sanjuro-client` I think would be pretty good names too. `game-core` I would like to see be split up into smaller domains and not just necessarily renamed.



> That is, if a user wants to run a bot, they will have to download an additional artifact. The install4j installers and the portable build will eventually no longer contain the code required to run a bot.

That is fair warning, I don't think we have a really clear support level defined beyond best effort. To have a server running for months on end takes some work, I believe the track record is that user bots are useful for handling surges/spikes, but are difficult to consider as part of permanent fleet. I'd be inclined for us to spend money on behalf of the community to spawn enough bot servers where we can handle usage spikes. @28 bots I think we may be looking pretty good, though it would be good to see about starting a track record of how full bots become.

**Comment by RoiEXLab** [comment]
@ssoloff 
> (Presumably, in the future, there will be a `game-headed` subproject that contains only the headed code, while all common code remains in `game-core`.)

True, however at some point we should split the game-core project again so the lobby only depends on the code in there it really needs.
For example: Both headed and headless servers depend on game parsing code which would probably remain in game-core. 
However the lobby doesn't need to know of this code, so we need another subproject.

File: game-core/build.gradle [hunk 1]
```diff
@@ -7,29 +7,13 @@ plugins {
 
 description = 'TripleA is a free online turn based strategy game and board game engine, similar to such board games as Axis & Allies or Risk.'
 mainClassName = 'games.strategy.engine.framework.GameRunner'
+version = getEngineVersion()
```
Comment (DanVanAtta) [inline game-core/build.gradle]: The extraction of `getEngineVersion` makes for a very clean update, I liked that it was very easy to re-use in another potential sub-project. While looking at version numbers, there were two things that came to mind:
1. Could the game-core project also use this same function?
2. Outside of the scope of this PR for a moment perhaps, but as a new artifact there is no strict requirement to have the version number in the artifact name. I noticed that the server does not have the version number built in, that was intentional when working with infrastructure scripting as the version number was already in the URL, and given the previous version may have already been downloaded, the extra information in the file name was ultimately just redundant/not helpful. I somewhat suspect when we update infrastructure, we'll find it could be a bit simpler to not have the version number in the jar file name. I'm happy for us to decide that a bit later when updating infrastructure scripting.

Comment (DanVanAtta) [reply game-core/build.gradle]: Ugh, I was mistaken in the above in some places. Perhaps we should unify this behavior. It was intended to stop adding version number to the artifact name as it was redundant.

But, it looks like we add version number.. For local build it is one artifact that gets 'dev' instead of the full engine version number, which is inconsistent and implies we have two mechanisms for doing this given that in the final release we have the full version number: 
```
./game-core/build/artifacts:
triplea-1.9.0.0.dev-all_platforms.zip  TripleA_1.9.0.0.dev_unix.sh	      TripleA_1.9.0.0.dev_windows-64bit.exe
TripleA_1.9.0.0.dev_macos.dmg	       TripleA_1.9.0.0.dev_windows-32bit.exe

./lobby/build/artifacts:
triplea-dev-server.zip
```
I think overall in this case the inconsistency and multiple ways of solving the same problem is the biggest issue beyond a redundant version number. That would imply that as a fast-follow we should use this new `getEngineVersion` method in each of the projects that produce a build artifact, or alternatively decide to drop version number from the build artifact name consistently. Long term I think that would be the best answer as it would lead to simplifications throughout the build and deploy pipeline. Getting user buy-in to not have the version number in the artifact name would take some work, but from a TripleA admin perspective we don't want version number in artifact names (deployment downloads by URL, version number is already baked in; and because the same version may already have been downloaded a possibly failed previous attempt, the version number does not necessarily help as we have to clean up any existing files anyways).

Sorry for the long paragraph there, I would suggest we plan to unify all artifacts to have version number in the same manner as a fast-follow; in parallel give notice we intend to remove version number from the game-core artifact, and then remove the version number simplifying the (gradle+travis) build and (infrastructure) deployment scripts.

**Comment by ssoloff** [comment]
@DanVanAtta said:

> `game-core` I would like to see be split up into smaller domains and not just necessarily renamed.

@RoiEXLab said:

> however at some point we should split the game-core project again so the lobby only depends on the code in there it really needs.

Sorry if it seemed that I implied that `game-core` would not be broken up further.  I'm sure it will.  It's just that in the short-term I see three `game-*` subprojects: one for the headless server, one for the headed client/server, and one for stuff common between them.  Those three then may be broken up further as we slowly decouple the code.

**Comment by ssoloff** [comment]
> Spending a few moments on naming I think will help us.

If I would have submitted this PR a few weeks ago, I would have named this subproject `game-server` and the forthcoming one `game-client`.  However, after thinking about it, I realized what we call the "headed client" is really both a client and a server.  I realize that's a lot of inside baseball, and may not be meaningful to anyone outside the dev team, but that's why I went with `game-headless` and `game-headed` to avoid potential confusion when it comes to naming packages.  For example, I'd hate to see a package name like `org.triplea.game.client.engine.server` or something similar.

With that said, maybe `game-server` (or `game-server-headless` for clarity) and `game-client` would be the best choices for now.  Then, if we decide to split the Swing and JFX clients, we would get `game-client-swing` and `game-client-javafx` (which would match nicely with the current package name for the JFX client).

>  These are all reasonable names:
>
>     game-headless
>     headless-server
>     bot-server
>     triplea-bot
>
> Similar:
>
>     game-headed
>     triplea-client
>     swing-client

1. I don't think `triplea` should be included in the subproject name.  Not only is it redundant, but it conflicts with the existing group name, which is already `triplea`.  So that if we tried to follow standard Maven/Gradle conventions to name our build outputs following the format `<group>-<artifact>-<version>-<classifier>.<ext>` (which I tried to do in this PR), we'd get a double `triplea` between `<group>` and `<artifact>`.
1. I think we should continue following the convention we've been using of starting with the more general concept and moving to the more specific in the subproject name.  For example, `client-swing` instead of `swing-client`.  That matches nicely with the top-level package name we'd presumably use in each subproject.

<hr>

**TL;DR:** I'm ok with switching the name of the project in this PR to either `game-server` or `game-server-headless` if that seems clearer.

File: game-core/build.gradle (hunk 1 above)
Comment (ssoloff) [reply game-core/build.gradle]: I think we should leave the version number in the artifact names.  A lot of times when debugging something, I will download multiple portable builds.  Having to manually name them during each download or renaming files that have been auto-suffixed by my browser (e,g, `file(1)`, `file(2)`, etc.) is annoying.

I haven't done anything with infrastructure, but for Travis, having the version in the file name hasn't really been an issue.  Usually the script already has the version in an environment variable, or simply using a bash globber when it doesn't know the version works fine if we're looking for something in the local file system (e.g. `java -jar bin/triplea-*-server-all.jar`).

**Comment by DanVanAtta** [comment]
@ssoloff for version number, I can go either way, it should be consistent though and we are currently all-over the map (migrations for example does not use one, server gets a 'dev' but not the full number, and game-core has its own copy of the engine number and soon we'll have a common method).  I do think there is some value in a pretty strict objective YAGNI, that if there is no user/business requirement, not to introduce one. I see benefits/cons in both directions, both in being flexibility but also in the benefits when taking a very hardline and simplifying like crazy (it does great things to code/support, and for this project that would seem called for in some ways). Regardless, I do not want that conversation/decision to hijack this PR, so to follow up I opened: https://github.com/triplea-game/triplea/issues/3957

That would leave naming, is the decision between: `game-headless ` or `game-server-headless` ? 
I would go for `game-headless` between the two personally.

**Comment by ssoloff** [comment]
Thanks for the feedback, everyone.  I'm removing the **&lt;PREVIEW>** label and will leave open for a bit in case there are any other comments regarding naming, etc.

**Comment by ssoloff** [comment]
Going to merge this now.  I'll follow-up with changes to infrastructure.
